test/session.cpp

```diff
@@ -0,0 +1,617 @@
+            mwr::usleep(5000);
+            sess.connect();
+        }
+    }
+
+    ~session_test() override {
+        sess.quit();
+        subp.terminate();
+    };
+
+    bool wait_for_target() {
+        constexpr int ms_timeout = 10000;
+        auto deadline = std::chrono::steady_clock::now() +
+                        std::chrono::milliseconds(ms_timeout);
+
+        while (std::chrono::steady_clock::now() < deadline) {
+            if (!sess.running())
+                return true;
+            mwr::usleep(1000);
+        }
+        return false;
+    }
+
+    const vector<u8> data1234{ 1, 2, 3, 4 };
+    const vector<u8> inf_loop_inst{ 0x00, 0x00, 0x00, 0x20 };
+
+    static constexpr u32 ADDR_IN_BOUND = 0x42;
+    static constexpr u32 ADDR_OO_BOUND = 0x8000;
+
+    session sess;
+
+    mwr::subprocess subp;
+};
+
+TEST_F(session_test, host_and_peer_data) {
+    EXPECT_TRUE(sess.is_connected());
+    EXPECT_STREQ(sess.host(), "localhost");
+    EXPECT_EQ(sess.port(), 4444);
+    ASSERT_NE(sess.peer(), nullptr);
+    EXPECT_GT(strlen(sess.peer()), 0);
+}
+
+TEST_F(session_test, connect_and_quit) {
+    // double connect
+    sess.connect();
+    EXPECT_TRUE(sess.is_connected());
+
+    sess.quit();
+    EXPECT_FALSE(sess.is_connected());
+
+    // double quit
+    sess.quit();
+    EXPECT_FALSE(sess.is_connected());
+}
+
+TEST_F(session_test, reconnect) {
+    sess.disconnect();
+    EXPECT_FALSE(sess.is_connected());
+
+    sess.connect();
+    EXPECT_TRUE(sess.is_connected());
+
+    sess.quit();
+    EXPECT_FALSE(sess.is_connected());
+
+    sess.connect();
+    EXPECT_FALSE(sess.is_connected());
+}
+
+TEST_F(session_test, dump) {
+    testing::internal::CaptureStdout();
+    sess.dump();
+    EXPECT_GT(testing::internal::GetCapturedStdout().size(), 0);
+}
+
+TEST_F(session_test, versions) {
+#ifdef GTEST_USES_SIMPLE_RE // windows simple regex: supports \d, not [] or {m}
+    constexpr auto sc_regex = R"(\d+\.\d+\.\d+)";
+    constexpr auto vcml_regex = R"(vcml-\d\d\d\d\.\d\d\.\d\d)";
+#else // POSIX ERE on linux: use classes and {m}
+    constexpr auto sc_regex = R"([0-9]+\.[0-9]+\.[0-9]+)";
+    constexpr auto vcml_regex = R"(vcml-[0-9]{4}\.[0-9]{2}\.[0-9]{2})";
+#endif
+
+    EXPECT_THAT(sess.sysc_version(), ContainsRegex(sc_regex));
+    EXPECT_THAT(sess.vcml_version(), ContainsRegex(vcml_regex));
+}
+
+TEST_F(session_test, time_and_cycles) {
+    EXPECT_EQ(sess.time_ns(), 0);
+    EXPECT_EQ(sess.cycle(), 1);
+}
+
+TEST_F(session_test, sessions) {
+    EXPECT_NE(session::get_sessions().size(), 0);
+}
+
+TEST_F(session_test, targets) {
+    auto& targets = sess.targets();
+    EXPECT_EQ(targets.size(), 1);
+
+    EXPECT_STREQ(targets.front().name(), "system.cpu");
+
+    target* targ;
+    targ = sess.find_target("system.cpu");
+    ASSERT_NE(targ, nullptr);
+
+    targ = sess.find_target("undefined-target");
+    ASSERT_EQ(targ, nullptr);
+}
+
+TEST_F(session_test, modules) {
+    module* mod;
```

Comment:
i would like to avoid `module`, since this will be a kwayword in c++20... can we do `vsp::module` here?